server/src/main/java/org/elasticsearch/search/fetch/FetchContext.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.fetch;
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.mapper.MapperService;
+import org.elasticsearch.index.query.ParsedQuery;
+import org.elasticsearch.search.SearchExtBuilder;
+import org.elasticsearch.search.fetch.subphase.FetchDocValuesContext;
+import org.elasticsearch.search.fetch.subphase.FetchFieldsContext;
+import org.elasticsearch.search.fetch.subphase.FetchSourceContext;
+import org.elasticsearch.search.fetch.subphase.FieldAndFormat;
+import org.elasticsearch.search.fetch.subphase.InnerHitsContext;
+import org.elasticsearch.search.fetch.subphase.ScriptFieldsContext;
+import org.elasticsearch.search.fetch.subphase.highlight.SearchHighlightContext;
+import org.elasticsearch.search.internal.ContextIndexSearcher;
+import org.elasticsearch.search.internal.SearchContext;
+import org.elasticsearch.search.rescore.RescoreContext;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Encapsulates state required to execute fetch phases
+ */
+public class FetchContext {
+
+    private FetchContext(SearchContext searchContext) {
+        this.searchContext = searchContext;
+    }
+
+    /**
+     * Create a FetchContext based on a SearchContext
+     */
+    public static FetchContext fromSearchContext(SearchContext context) {
+        return new FetchContext(context);
+    }
+
+    private final SearchContext searchContext;
```

Comment:
Could you float it to the top so it doesn't get lost? I get really confused if member variables aren't above member methods. I'm just super used to that.